### ISSUE TITLE
S3 bucket inventory

### ISSUE THREAD

**Issue by lason-ovh** [ovh/public-cloud-roadmap#320]
Bucket inventory is one the tools that can help you manage your storage. It provides an asynchronous alternative to "ListObjects" that allows you to combine the results with other S3 functionalities such as S3 select to perform powerful analytic tasks.

In practical terms, bucket inventory generates on a daily or weekly basis a csv file that lists your objects and their corresponding metadata such as size, storage class, and so on.

One typical use case is to use Bucket inventory to generate a csv listing and then use S3 select to query that file with SQL to get the total size of the bucket, total number of objects, % of objects by storage class, ...etc.

What we will provide is an API to upload, download and delete a configuration file that tells the Object Storage:
- what metadata to export
- where to send the results
- how often to generate the inventory

The following configuration elements will be managed:
- object metadata selection
- destination bucket
- filters on objects
- schedule

The following configuration elements will not be managed for now:
- encryption of csv export in destination bucket
- export to a bucket owner by a different account
- export to json and Apache parquet format

**Comment by omascia** [comment]
Hi! Questions.
1) Do you intend to implement Select as Amazon S3 Select? I guess that's the clear intend, but better be sure.
https://docs.aws.amazon.com/AmazonS3/latest/userguide/selecting-content-from-objects.html
2) Will this bucket inventory be (even partially) compatible with Amazon S3 Inventory?
https://docs.aws.amazon.com/AmazonS3/latest/userguide/storage-inventory.html
So are we talking here clearly about S3:PutInventoryConfiguration and similar configuration descriptions and more importantly similar generated report files contents?

If yes on all the lines, even though some details would not be supported, but nothing supported would be implemented differently, forcing alternate code paths for S3 "à la OVH" or S3 "à la Amazon"... then I will plus-plus-plus this feature and upvote here to the sky-limit.

**Comment by lason-ovh** [comment]
yes and yes
the idea is to provide our dear customers a combination of the 2 to enable not only analytical features on their bucket but also analytical features on any given set of compatible data

**Comment by omascia** [comment]
Can't wait for it! ;)

**Comment by polmonso** [comment]
Does the acknowledgement include select-object-content ? to query files inplace?

**Comment by lason-ovh** [comment]
> Does the acknowledgement include select-object-content ? to query files inplace?

Yes, initially, we intended to implement S3 Select after this feature (and thus support select-object-content). However, S3 Select has since been deprecated and the feature is no more available for new customers. In order to maximize our compatibility with the S3 API, the chances we implement S3 Select have since then become very slim.

**Comment by welyb-gragueneau** [comment]
Any updates on this feature? It would be very useful for us as well — we have huge buckets (over 20 TB) and cannot properly monitor them using list object operations, as that would take too long.

https://docs.aws.amazon.com/fr_fr/AmazonS3/latest/userguide/configure-inventory.html
